src/mongoose_log_filter.erl

```diff
@@ -0,0 +1,125 @@
+    BinPacket = exml_packet:to_binary(Packet),
+    Msg2 = maps:put(packet, BinPacket, maps:remove(exml_packet, Msg)),
+    Event#{msg => {report, Msg2}};
+format_packet_filter(Event, _) ->
+    Event.
+
+format_stacktrace_filter(Event=#{msg := {report, Msg=#{stacktrace := S}}}, _) ->
+    Event#{msg => {report, Msg#{stacktrace => format_stacktrace(S)} }};
+format_stacktrace_filter(Event, _) ->
+    Event.
+
+format_acc(#{origin_pid := OriginPid, timestamp := TS, stanza := StanzaMap}) ->
+    Map = format_stanza_map(StanzaMap),
+    Map#{acc_timestamp => format_microseconds(TS),
+         origin_pid => format_term(OriginPid)};
+format_acc(_) ->
+    #{}.
+
+format_stanza_map(#{element := Elem, from_jid := From, to_jid := To}) ->
+    #{packet => exml:to_binary(Elem),
+      from_jid => jid:to_binary(From),
+      to_jid => jid:to_binary(To)};
+format_stanza_map(_) ->
+    #{}.
+
+preserve_acc_filter(Event=#{msg := {report, Msg=#{acc := Acc}}}, _) ->
+    Event#{msg => {report, Msg#{acc_original => format_term(Acc)}}};
+preserve_acc_filter(Event, _) ->
+    Event.
+
+
+c2s_state_to_map(#state{socket = Socket, streamid = StreamId,
+                        jid = Jid, sid = Sid}) ->
+    SocketMap = ejabberd_socket:format_socket(Socket),
+    SocketMap#{
+      streamid => StreamId,
+      jid => maybe_jid_to_binary(Jid),
+      user => maybe_jid_to_luser(Jid),
+      server => maybe_jid_to_lserver(Jid),
+      resource => maybe_jid_to_lresource(Jid),
+      session_started => maybe_sid_to_timestamp(Sid)}.
+
+format_term(X) -> iolist_to_binary(io_lib:format("~0p", [X])).
+
+maybe_jid_to_binary(Jid = #jid{}) -> jid:to_binary(Jid);
+maybe_jid_to_binary(_) -> undefined.
+
+maybe_jid_to_luser(#jid{luser = LUser}) -> LUser;
+maybe_jid_to_luser(_) -> undefined.
+
+maybe_jid_to_lserver(#jid{lserver = LServer}) -> LServer;
+maybe_jid_to_lserver(_) -> undefined.
+
+maybe_jid_to_lresource(#jid{lresource = LResource}) -> LResource;
+maybe_jid_to_lresource(_) -> undefined.
+
+maybe_sid_to_timestamp({Timestamp, _Pid}) -> format_microseconds(Timestamp);
+maybe_sid_to_timestamp(_) -> undefined.
+
+format_microseconds(N) ->
+    calendar:system_time_to_rfc3339(N, [{unit, microsecond},
+                                        {offset, 0},
+                                        {time_designator, $T}]).
+
+
+format_stacktrace(Stacktrace) ->
+    iolist_to_binary(do_format_stacktrace(Stacktrace)).
+
+do_format_stacktrace([{Mod,Fun,Arity,Info}|T]) ->
+    Line = proplists:get_value(line, Info, 0),
+    H = io_lib:format("~p:~p/~p:~p", [Mod, Fun, Arity, Line]),
+    more_format_stacktrace(H, T);
+do_format_stacktrace([Other|T]) ->
+    H = io_lib:format("~p", [Other]),
+    more_format_stacktrace(H, T);
+do_format_stacktrace([]) ->
+    [].
+
+more_format_stacktrace(H, []) ->
+    [H];
+more_format_stacktrace(H, T) ->
+    [H, " "|do_format_stacktrace(T)].
+
+filter_undefined(Map) ->
+    maps:filter(fun(K, _) -> K =/= undefined end, Map).
```

Comment:
I think that we want to filter out `undefined` _values_, not _keys_, right? 🤔

Reply:
good catch